File: python/promethean/corridor_generator.py
import math
from typing import List

from promethean.point import Point
from promethean.room_generator import Room, RoomGenerator
from promethean.pseudo_random import PseudoRandom
from promethean.options import PathFinderTile, PathFinderOptions, RoomType, Options
from promethean.path_finder import PathFinder

class Corridor:
    def __init__(self):
        self._tiles: List[Point] = []

    def __repr__(self) -> str:
        return str([str(p) for p in self._tiles])


def calculate_distance_between_2_points(origin: Point, point: Point) -> float:
    xa_minux_xb_squared: int = (point._x - origin._x) ** 2
    ya_minus_yb_squared: int = (point._y - origin._y) ** 2

    return math.sqrt(xa_minux_xb_squared + ya_minus_yb_squared)


def room_distance_from_origin_comparer(room1: Room, room2: Room) -> int:
    reference = Point(0, 0)
    room1_distance_from_reference = calculate_distance_between_2_points(reference, room1._room_center)
    room2_distance_from_reference = calculate_distance_between_2_points(reference, room2._room_center)
    if room1_distance_from_reference < room2_distance_from_reference:
        return -1
    elif room1_distance_from_reference > room2_distance_from_reference:
        return 1
    else:
        return 0


def room_distance_from_origin_key(room: Room) -> float:
    reference = Point(0, 0)
    return calculate_distance_between_2_points(reference, room._room_center)


def generate_pathing_grid(rooms: List[Room], options: Options) -> List[List[PathFinderTile]]:
    pathable_level: List[List[PathFinderTile]] = [[PathFinderTile.Blocked for j in range(options.level_width)]for i in range(options.level_height)]

    x: int = 0
    y: int = 0
    for x in range(options.border, options.level_height - options.border):
        for y in range(options.border, options.level_width - options.border):
            pathable_level[x][y] = PathFinderTile.Pathable

    for room in rooms:
        for x_offset in range(0 - options.room_border, room._height + options.room_border):
            for y_offset in range(0 - options.room_border, room._width + options.room_border):
                x = room._position._x + x_offset
                y = room._position._y + y_offset

                if x < options.border or x >= options.level_height - options.border:
                    continue
                if y < options.border or y >= options.level_width - options.border:
                    continue

                if x == room._room_center._x or y == room._room_center._y:
                    pathable_level[x][y] = PathFinderTile.Pathable
                    continue

                pathable_level[x][y] = PathFinderTile.Blocked

    return pathable_level


def generate_corridors(rooms: List[Room], options: Options) -> List[Corridor]:
    pathable_level: List[List[PathFinderTile]] = generate_pathing_grid(rooms, options)

    pathfinder: PathFinder = PathFinder(pathable_level, PathFinderOptions())
    rooms.sort(key=room_distance_from_origin_key)

    corridors: List[Corridor] = []
    for index in range(0, len(rooms) - 1):
        current_room: Room = rooms[index]
        next_room: Room = rooms[index + 1]

        path = pathfinder.find_path(current_room._room_center, next_room._room_center)

        corridor_length: int = len(path)
        if corridor_length == 0:
            # if the path is empty, then we fails to connect two rooms by a corridor
            # in this case the level is not connected
            # we should change generator parameters or change the random seed
            continue

        corridor = Corridor()
        for i, p in enumerate(path):
            corridor._tiles.append(p)
            # add non-pathable tile in the corridor corners
            if i > 0 and i < corridor_length - 1:
                prev_p: Point = path[i - 1]
                next_p: Point = path[i + 1]
                x: int = p._x
                y: int = p._y
                prev_x: int = prev_p._x
                prev_y: int = prev_p._y
                next_x: int = next_p._x
                next_y: int = next_p._y
                if prev_x == x and x != next_x:
                    if prev_y < y:
                        if x < next_x:
                            pathfinder.block_point(Point(x, y + 1))
                        else:
                            pathfinder.block_point(Point(x + 1, y + 1))
                    else:
                        if x < next_x:
                            pathfinder.block_point(Point(x - 1, y - 1))
                        else:
                            pathfinder.block_point(Point(x - 1, y + 1))
                elif prev_y == y and y != next_y:
                    if prev_x < x:
                        if y < next_y:
                            pathfinder.block_point(Point(x + 1, y - 1))
                        else:
                            pathfinder.block_point(Point(x + 1, y + 1))
                    else:
                        if y < next_y:
                            pathfinder.block_point(Point(x - 1, y - 1))
                        else:
                            pathfinder.block_point(Point(x, y + 1))

        corridors.append(corridor)

    return corridors

File: rust/promethean/src/corridor_generator.rs
#[path = "path_finder.rs"] mod path_finder;

use crate::level::point::Point;
use crate::level::options::{PathFinderTile, Options};
use crate::level_generator::room_generator::{Room};
use path_finder::PathFinder;

use std::cmp::Ordering;

#[derive(Debug)]
pub struct Corridor {
    tiles: Vec<Point>
}

impl Corridor {
    pub fn new(points: Vec<Point>) -> Corridor {
        return Corridor { tiles: points };
    }

    pub fn get_tiles(&self) -> &Vec<Point> {
        return &self.tiles;
    }
}

fn calculate_distance_between_2_points(origin: &Point, point: &Point) -> f64 {
    let xa_minux_xb_squared: i32 = (point.x() - origin.x()) * (point.x() - origin.x());
    let ya_minus_yb_squared: i32 = (point.y() - origin.y()) * (point.y() - origin.y());

    return ((xa_minux_xb_squared + ya_minus_yb_squared) as f64).sqrt();
}

pub fn generate_pathing_grid(rooms: &Vec<Room>, options: &Options) -> Vec<Vec<PathFinderTile>> {
    let mut pathable_level: Vec<Vec<PathFinderTile>> = Vec::with_capacity(options.level_height());
    for x in 0..options.level_height() {
        let mut x_array: Vec<PathFinderTile> = Vec::with_capacity(options.level_width());
        for y in 0..options.level_width() {
            if x < options.border() || x >= options.level_height() - options.border() || y < options.border() || y >= options.level_width() - options.border() {
                x_array.push(PathFinderTile::Blocked);
            }
            else {
                x_array.push(PathFinderTile::Pathable);
            }
        }
        pathable_level.push(x_array);
    }

    let mut x: usize;
    let mut y: usize;

    for r_index in 0..rooms.len() {
        let room = &rooms[r_index];
        for x_offset in -(options.room_border() as i32)..((room.height() + options.room_border()) as i32) {
            for y_offset in -(options.room_border() as i32)..((room.width() + options.room_border()) as i32) {
                x = (room.position().x() + x_offset) as usize;
                y = (room.position().y() + y_offset) as usize;

                if x < options.border() || x >= options.level_height() - options.border() {
                    continue;
                }
                if y < options.border() || y >= options.level_width() - options.border() {
                    continue;
                }
                
                let room_center = room.center();
                if x == room_center.x() as usize || y == room_center.y() as usize {
                    pathable_level[x][y] = PathFinderTile::Pathable;
                    continue;
                }

                pathable_level[x][y] = PathFinderTile::Blocked;
            }
        }
    }

    return pathable_level;
}

pub fn generate_corridors(rooms: &mut Vec<Room>, options: &Options) -> Vec<Corridor> {
    if rooms.len() <= 1 {
        return Vec::new();
    }

    let pathable_level: Vec<Vec<PathFinderTile>> = generate_pathing_grid(rooms, options);

    let mut pathfinder: PathFinder = PathFinder::new(pathable_level, 2000);
    rooms.sort_by(|room1, room2| {
        let reference = Point::new(0, 0);
        let room1_distance_from_reference = calculate_distance_between_2_points(&reference, room1.center());
        let room2_distance_from_reference = calculate_distance_between_2_points(&reference, room2.center());
        if room1_distance_from_reference < room2_distance_from_reference {
            return Ordering::Less;
        }
        else if room1_distance_from_reference > room2_distance_from_reference {
            return Ordering::Greater;
        }
        else{
            return Ordering::Equal;
        }
    });

    let mut corridors = Vec::with_capacity(rooms.len() - 1);
    for index in 0..rooms.len() - 1 {
        let current_room = &rooms[index];
        let next_room = &rooms[index + 1];

        let path: Vec<Point> = pathfinder.find_path(current_room.center(), next_room.center());

        let corridor_length = path.len();
        if corridor_length == 0 {
            continue;
        }

        for i in 1..corridor_length - 1 {
            let p = &path[i];
            let prev_p = &path[i - 1];
            let next_p = &path[i + 1];
            let x = p.x();
            let y = p.y();
            let prev_x: i32 = prev_p.x();
            let prev_y: i32 = prev_p.y();
            let next_x: i32 = next_p.x();
            let next_y: i32 = next_p.y();
            if prev_x == x && x != next_x {
                if prev_y < y {
                    if x < next_x {
                        pathfinder.block_point(Point::new(x, y + 1));
                    }
                    else {
                        pathfinder.block_point(Point::new(x + 1, y + 1));
                    }
                }
                else {
                    if x < next_x {
                        pathfinder.block_point(Point::new(x - 1, y - 1));
                    }
                    else {
                        pathfinder.block_point(Point::new(x - 1, y + 1));
                    }
                }
            }
            else if prev_y == y && y != next_y {
                if prev_x < x {
                    if y < next_y {
                        pathfinder.block_point(Point::new(x + 1, y - 1));
                    }
                    else {
                        pathfinder.block_point(Point::new(x + 1, y + 1));
                    }
                }
                else {
                    if y < next_y {
                        pathfinder.block_point(Point::new(x - 1, y - 1));
                    }
                    else {
                        pathfinder.block_point(Point::new(x, y + 1));
                    }
                }
            }
        }
        let corridor = Corridor::new(path);

        corridors.push(corridor);
    }

    return corridors;
}